include/up-cpp/utils/CallbackConnection.h

```diff
@@ -200,14 +304,97 @@ struct [[nodiscard]] CalleeHandle {
+		return locked_connection && (*locked_connection);
+	}
 
 private:
-	std::shared_ptr<Conn> connection_;
+	std::weak_ptr<Conn> connection_;
 	std::shared_ptr<Callback> callback_;
 	std::optional<Cleanup> cleanup_;
 };
 
+/// @brief Thrown if a default constructed or reset() CallerHandle is called.
+///
+/// This does not occur if the connection was broken from the Callee end. When
+/// this is thrown, it represents a programmatic error around the use of a
+/// CallerHandle that needs to be corrected.
+struct BadCallerAccess : public std::logic_error {
+	template <typename... Args>
+	BadCallerAccess(Args&&... args)
+	    : std::logic_error(std::forward<Args>(args)...) {}
+};
+
+template <typename RT, typename... Args>
+struct [[nodiscard]] CallerHandle {
+	using Conn = Connection<RT, Args...>;
+
+	/// @brief Default construction results in an non-connected handle
+	CallerHandle() = default;
+
+	/// @brief Creates a connected handle. Only usable by Connection
+	CallerHandle(std::shared_ptr<Conn> connection,
+	             typename Conn::PrivateConstructToken)
+	    : connection_(connection) {
+		if (!connection_) {
+			throw BadConnection(
+			    "Attempted to create a connected CallerHandle with bad "
+			    "connection pointer");
+		}
+	}
+
+	/// @brief Drops this instance's copy of the handle, severing the
+	///        connection if no other CallerHandles represent the same handle.
+	void reset() { connection_.reset(); }
+
+	/// @brief Check if the connection is still valid.
+	///
+	/// @returns
+	///     * True if the connection is valid (i.e. at least one other
+	///       reference to the connection still exists).
+	///     * False if the connection has been broken (i.e. This handle has
+	///       been reset/moved, or all other references to the connection
+	///       have been discarded)
+	explicit operator bool() const { return connection_ && (*connection_); }
+
+	/// @throws BadCallerAccess if this handle has been default constructed OR
+	///         reset() has left it without a valid conneciton pointer.
+	auto operator()(Args&&... args) {
+		if (!connection_) {
+			// Note: this only occurs with default constructed CallerHandle or
+			// when CallerHandle::reset() has been called on a given instance.
+			// This would represent a programmatic error around the use of a
+			// CallerHandle that needs to be corrected.
+			//
+			// This will not be thrown when the connection has been broken from
+			// the CalleeHandle side of the connection as that could introduce
+			// race conditions and is outside of the control of the code
+			// holding the CallerHandle..
+			throw BadCallerAccess(
+			    "Cannot call a CallerHandle that is in the reset state");
+		}
+		return (*connection_)(std::forward<Args>(args)...);
+	}
+
+	/// @brief Comparison based on connection pointer value
+	///
+	/// Allows `CallerHandle` to be stored in a container that sorts on value.
+	auto operator<(const CallerHandle& other) const {
```

Comment:
Is the point of this to allow sorted lists by caller address?  This doesn't seem to provide any value.

Reply:
It allows for handles to be stored in a `std::map` if needed for reverse lookups during the cleanup operation